src/Compiler/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs

```diff
@@ -544,6 +544,58 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument codeDocument)
         }
     }
 
+    protected void AssertHtmlSourceMappingsMatchBaseline(RazorCodeDocument codeDocument)
+    {
+        if (FileName == null)
+        {
+            var message = $"{nameof(AssertHtmlSourceMappingsMatchBaseline)} should only be called from an integration test ({nameof(FileName)} is null).";
+            throw new InvalidOperationException(message);
+        }
+
+        var htmlDocument = codeDocument.GetHtmlDocument();
+        Assert.NotNull(htmlDocument);
+
+        var baselineFileName = Path.ChangeExtension(FileName, ".html.mappings.txt");
+        var serializedMappings = SourceMappingsSerializer.Serialize(htmlDocument, codeDocument.Source);
+
+        if (GenerateBaselines)
+        {
+            var baselineFullPath = Path.Combine(TestProjectRoot, baselineFileName);
+            File.WriteAllText(baselineFullPath, serializedMappings);
+            return;
+        }
+
+        var testFile = TestFile.Create(baselineFileName, GetType().GetTypeInfo().Assembly);
+        if (!testFile.Exists())
+        {
+            throw new XunitException($"The resource {baselineFileName} was not found.");
+        }
+
+        var baseline = testFile.ReadAllText();
+
+        // Normalize newlines to match those in the baseline.
+        var actualBaseline = serializedMappings.Replace("\r", "").Replace("\n", "\r\n");
+
+        Assert.Equal(baseline, actualBaseline);
```

Comment:
`AssertEx.AssertEqualToleratingWhitespaceDifferences`. It'll give a much better error message too.